Custom domains documentation (#739)
fermyon/developer: content/cloud/custom-domains-tutorial.md

* A Spin application you would like to apply the custom domain to. If you aren't familiar with how to deploy an application on Fermyon Cloud, check out our [develop a Spin application](./develop.md) article to learn how to build a Spin application and then our [deploy a Spin application article](./deploy.md) for deployment instructions. 

## Select Your Spin Application

In the [Fermyon Cloud Dashboard](https://cloud.fermyon.com), navigate to your Spin application of choice and click the "Manage Domains" button.

![DRAFT SCREENSHOT - app panel view](/static/image/custom-domains/app-panel-view.png)

## Input Your Custom Domain

Select the "Add custom domain" button and input your desired domain name. 

![DRAFT SCREENSHOT - step 1](/static/image/custom-domains/Step-1.png)

Please add the complete domain name (i.e., `www.example.com`) as Fermyon Cloud does not support redirects at this time. Click save to confirm you would like to add this domain name to Fermyon Cloud.

![DRAFT SCREENSHOT - step 2](/static/image/custom-domains/Step-2.png)

Up next, you'll need to configure your DNS records with your registrar to point at Fermyon's name servers (NS). 

> >>If you click "I'll do this later" you will be pausing the workflow and your *custom domain will not work yet* with your Spin application.

Author: Good callout - I'll add a sentence to clarify how the user can resume their journey.

> >>If you click "I'll do this later" you will be pausing the workflow and your *custom domain will not work yet* with your Spin application.

Reviewer: I like this extra sentence Kenzie.

![DRAFT SCREENSHOT - step 3](/static/image/custom-domains/Step-3.png)

## Update Your Domain Registrar With Fermyon Cloud Name Servers

> If you've never delegated this domain to Fermyon Cloud before, you'll proceed strate to updating your nameserver records with your domain registrar with the nameservers generated by Fermyon Cloud. Each domain registrar has it's own unqiue set of instructions for completing this workflow. Please visit your domain registrar's documentation for instructions. 

Reviewer: Thanks, fixed these up in latest push.

> If you've never delegated this domain to Fermyon Cloud before, you'll proceed strate to updating your nameserver records with your domain registrar with the nameservers generated by Fermyon Cloud. Each domain registrar has it's own unqiue set of instructions for completing this workflow. Please visit your domain registrar's documentation for instructions. 

Author: Thank you for the feedback! I took another stab at it. Let me know if this helps clarify the step.

> If you've never delegated this domain to Fermyon Cloud before, you'll proceed strate to updating your nameserver records with your domain registrar with the nameservers generated by Fermyon Cloud. Each domain registrar has it's own unqiue set of instructions for completing this workflow. Please visit your domain registrar's documentation for instructions. 

Reviewer: Nice!

![DRAFT SCREENSHOT - step 4](/static/image/custom-domains/Step-4.png)

> In addition to generating NS records, Fermyon Cloud will generate a Let's Encrypt certificate on your behalf. 

Reviewer 2: This slightly mixes two kinds of "generate" - there's some stuff on the screen that the reader needs to go take to their registrar, and an internal that FC stashes away.  Like reading this I wasn't sure if I needed to do anything with the cert?

I'd consider removing the mention of the cert here, and when all is set up, remark that FC has generated (and will keep renewing) a cert for you.  Or say something like: "Note that you do _not_ need to provide a SSL certificate.  Fermyon Cloud creates one internally, and will keep renewing it for the lifetime of your application."  But I guess that applies equally to the "repeat delegation" case?

Author: I created a dedicated "callout" about cert management! I can move it all the way to the bottom if you think that would make more sense, but give it a read! :) I agree, the way it was previously written was confusing

Reviewer: Feels great to have that sentence about Fermyon taking care of SSL renewal for the life of the app. That's a huge deal.
This is a GREAT value add.

### Repeat Delegation Only: Update Domain Registrar with TXT Record

> If you've delegate the same domain before with Fermyon Cloud, you will be prompted to add a TXT record as well as part of the verification process. 

Reviewer 2: As well as what - is this hinting that the NS step applies here as well?

Author: Yes it is, but we are removing TXT validation from MVP so this section is no longer valid

> If you've delegate the same domain before with Fermyon Cloud, you will be prompted to add a TXT record as well as part of the verification process. 

Reviewer 2: ```suggestion
If you've delegated the same domain before with Fermyon Cloud, you will be prompted to add a TXT record as well as part of the verification process. 
```

Reviewer: Thanks; fixed in push.

![DRAFT SCREENSHOT - step 4.1](/static/image/custom-domains/Step-4.1.png)

## Verify domain has been delegated

Please wait for up to 72 hours for the change to process. Once the verifcation process has completed, you should see a green checkmark next to your custom domain. 

> ![DRAFT SCREENSHOT - step 4.1](/static/image/custom-domains/Step-5.png)

Reviewer 2: ```suggestion
![DRAFT SCREENSHOT - step 5](/static/image/custom-domains/Step-5.png)
```

Reviewer: Nice catch, fixed in next push.

- [Verify Domain Has Been Delegated](#verify-domain-has-been-delegated)
- [Next Steps](#next-steps)

Custom domains enable you to bring your own domain to Fermyon Cloud.  Unlike generic or randomly generated domains, a custom domain reflects the brand, identity, or purpose of the workload. For a conceptual overview of custom domains on Fermyon Cloud, visit (Custom Domains on Fermyon Cloud)[./custom-domain.md]. This tutorial will walk you through how to apply a custom domain to your Spin application running on Fermyon Cloud. At the end of this tutorial, you will have delegated your domain to Fermyon Cloud. Let's get started!

## Prerequisites

To ensure the tutorial goes smoothly, please ensure you have the following: 

* Most importantly, you must have a domain you own. If you don't have a domain, please visit a domain registrar like [Namecheap](https://www.namecheap.com/) or [Google Domains](https://domains.google) to purchase one. 
* A Fermyon Cloud account that is set up via your preferred [GitHub user account](https://docs.github.com/account-and-profile/setting-up-and-managing-your-personal-account-on-github/managing-email-preferences/remembering-your-github-username-or-email)

> * A Spin application you would like to apply the custom domain to. If you aren't familiar with how to deploy an application on Fermyon Cloud, check out our [develop a Spin application](./develop.md) article to learn how to build a Spin application and then our [deploy a Spin application article](./deploy.md) for deployment instructions. 

Reviewer 2: Maybe clarify that this is a _deployed_ application.  "Spin application" could be local.

```suggestion
* A Spin application already deployed to Fermyon Cloud that you would like to apply the custom domain to. If you aren't familiar with how to deploy an application on Fermyon Cloud, check out our [develop a Spin application](./develop.md) article to learn how to build a Spin application and then our [deploy a Spin application article](./deploy.md) for deployment instructions. 
```

## Select Your Spin Application

In the [Fermyon Cloud Dashboard](https://cloud.fermyon.com), navigate to your Spin application of choice and click the "Manage Domains" button.

![DRAFT SCREENSHOT - app panel view](/static/image/custom-domains/app-panel-view.png)

## Input Your Custom Domain

Select the "Add custom domain" button and input your desired domain name. 

![DRAFT SCREENSHOT - step 1](/static/image/custom-domains/Step-1.png)

> Please add the complete domain name (i.e., `www.example.com`) as Fermyon Cloud does not support redirects at this time. Click save to confirm you would like to add this domain name to Fermyon Cloud.

Reviewer 2: ```suggestion
Please add the complete domain name (e.g. `www.example.com`) as Fermyon Cloud does not support redirects at this time. Click save to confirm you would like to add this domain name to Fermyon Cloud.
```

my inner pedant and my outer pedant are in complete solidarity on this one

![DRAFT SCREENSHOT - step 2](/static/image/custom-domains/Step-2.png)

Up next, you'll need to configure your DNS records with your registrar to point at Fermyon's name servers (NS). 

>>If you click "I'll do this later" you will be pausing the workflow and your *custom domain will not work yet* with your Spin application. You can navigate back to the "Manage Domains" workflow to resume your journey. 

![DRAFT SCREENSHOT - step 3](/static/image/custom-domains/Step-3.png)

## Update Your Domain Registrar With Fermyon Cloud Name Servers

> At this step, Fermyon Cloud will check to see if you've previously associated your domain with Fermyon Cloud before. If this is the first time Fermyon Cloud has seen the domain, you will need to visit your domain registrar to delegate your domain to Fermyon Cloud. Domain delegation is the process of reassigning nameservers for a domain name to another DNS provider, which allows said nameservers to handle all DNS requests and queries for that domain. 

Reviewer 2: So if FC _has_ seen the domain before then I don't need to do anything?  The "in this step" paragraph doesn't really elaborate on this.
```suggestion
At this point, Fermyon Cloud checks to see if you've previously associated your domain with Fermyon Cloud before.

If you _have_ previously associated your domain, there's nothing more to do in this step!  Skip ahead to the next step.

If, however, this is the first time Fermyon Cloud has seen the domain, you will need to visit your domain registrar to delegate your domain to Fermyon Cloud. Domain delegation is the process of reassigning nameservers for a domain name to another DNS provider, which allows said nameservers to handle all DNS requests and queries for that domain. 
```

(not sure I've brought out the contrast, and it would be better if the "not done before" flow was the main one... but hopefully gives you ideas)

Reviewer: I love this re-jigging. Much clearer.

> In this step, you will update your existing nameserver records at your domain registrar to be replaced by Fermyon Cloud's namesever records. Fermyon Cloud will automatically generate four nameserver records for you. 

Reviewer 2: nit: semi-repeat "At this step" / "In this step".
```suggestion
In this step, you will update your existing nameserver records at your domain registrar to be replaced by Fermyon Cloud's nameserver records. Fermyon Cloud will automatically generate four nameserver records for you. 
```

![DRAFT SCREENSHOT - step 4](/static/image/custom-domains/Step-4.png)

Please copy all four nameserver records and use them to replace your name server records with your domain registrar. Each domain registrar has its own unique set of instructions for completing this workflow. Please visit your domain registrar's documentation for instructions on how to update your nameserver records. For your convenience, we've sourced instructions for [Google Domains](https://support.google.com/domains/answer/3290309?hl=en#) and [Namecheap](https://www.namecheap.com/support/knowledgebase/article.aspx/767/10/how-to-change-dns-for-a-domain/).

>> Note that you do not need to provide a SSL certificate. Fermyon Cloud creates one internally using Let's Encrypt, and will keep renewing it for the lifetime of your Spin application. 

## Verify Domain Has Been Delegated

> Please wait for up to 72 hours for the process change. Once the verification process has been completed, you should see a green checkmark next to your custom domain. 

Reviewer 2: Not sure if "process change" is a typo... "for the change to be processed"? "to process the change"?

Reviewer: I like "for the change to be processed".

> Please wait for up to 72 hours for the process change. Once the verification process has been completed, you should see a green checkmark next to your custom domain. 

Reviewer 2: Is the "verification process" the process that takes 72 hours?

title = "Apply Custom Domain"

> date = "2023-07-26T12:00:00Z"

Reviewer: ```suggestion
date = "2023-07-26T16:00:00Z"
```

template = "cloud_main"
enable_shortcodes = true
[extra]
url = "https://github.com/fermyon/developer/blob/main/content/cloud/custom-domains-tutorial.md"

---

- [Prerequisites](#prerequisites)
- [Select Your Spin Application](#select-your-spin-application)
- [Input Your Custom Domain](#input-your-custom-domain)
- [Update Your Domain Registrar With Fermyon Cloud Name Servers](#update-your-domain-registrar-with-fermyon-cloud-name-servers)
- [Verify Domain Has Been Delegated](#verify-domain-has-been-delegated)
- [Next Steps](#next-steps)

Custom domains enable you to bring your own domain to Fermyon Cloud.  Unlike generic or randomly generated domains, a custom domain reflects the brand, identity, or purpose of the workload. For a conceptual overview of custom domains on Fermyon Cloud, visit (Custom Domains on Fermyon Cloud)[./custom-domain.md]. This tutorial will walk you through how to apply a custom domain to your Spin application running on Fermyon Cloud. At the end of this tutorial, you will have delegated your domain to Fermyon Cloud. Let's get started!

## Prerequisites

To ensure the tutorial goes smoothly, please ensure you have the following: 

> * Most importantly, you must have a domain you own. If you don't have a domain, please visit a domain registrar like [Namecheap](https://www.namecheap.com/) or [Google Domains](https://domains.google) to purchase one. 

Reviewer: ```suggestion
* Most importantly, you must have a domain you own. If you don't have a domain, please visit a domain registrar like [Namecheap](https://www.namecheap.com/) or [GoDaddy](https://www.godaddy.com/) to purchase one. 
```

In this step, you will update your existing nameserver records at your domain registrar to be replaced by Fermyon Cloud's nameserver records. Fermyon Cloud will automatically generate four nameserver records for you. 

![DRAFT SCREENSHOT - step 4](/static/image/custom-domains/Step-4.png)

Please copy all four nameserver records and use them to replace your name server records with your domain registrar. Each domain registrar has its own unique set of instructions for completing this workflow. Please visit your domain registrar's documentation for instructions on how to update your nameserver records. For your convenience, we've sourced instructions for [Google Domains](https://support.google.com/domains/answer/3290309?hl=en#) and [Namecheap](https://www.namecheap.com/support/knowledgebase/article.aspx/767/10/how-to-change-dns-for-a-domain/).

>> Note that you do not need to provide a SSL certificate. Fermyon Cloud creates one internally using Let's Encrypt, and will keep renewing it for the lifetime of your Spin application. 

## Verify Domain Has Been Delegated

> Please wait for up to 72 hours for the process change. Once the verification process has been completed, you should see a green checkmark next to your custom domain. 

Reviewer: ```suggestion
Please wait for up to 72 hours for the change to be processed. Once the verification process has been completed, you should see a green checkmark next to your custom domain. 
```
